src/absil/bytes.fs

```diff
@@ -26,10 +33,290 @@ module internal Bytes =
         Array.append (System.Text.Encoding.UTF8.GetBytes s) (ofInt32Array [| 0x0 |]) 
 
     let stringAsUnicodeNullTerminated (s:string) = 
-        Array.append (System.Text.Encoding.Unicode.GetBytes s) (ofInt32Array [| 0x0;0x0 |]) 
+        Array.append (System.Text.Encoding.Unicode.GetBytes s) (ofInt32Array [| 0x0;0x0 |])
+
+[<AbstractClass>]
+type ByteMemory () =
+
+    abstract Item: int -> byte with get, set
+
+    abstract Length: int
+
+    abstract ReadBytes: pos: int * count: int -> byte[]
+
+    abstract ReadInt32: pos: int -> int
+
+    abstract ReadUInt16: pos: int -> uint16
+
+    abstract ReadUtf8String: pos: int * count: int -> string
+
+    abstract Slice: pos: int * count: int -> ByteMemory
+
+    abstract CopyTo: Stream -> unit
+
+    abstract Copy: srcOffset: int * dest: byte[] * destOffset: int * count: int -> unit
+
+    abstract ToArray: unit -> byte[]
+
+    abstract AsStream: unit -> Stream
+
+    abstract AsReadOnlyStream: unit -> Stream
+
+[<Sealed>]
+type ByteArrayMemory(bytes: byte[], offset, length) =
+    inherit ByteMemory()
+
+    do
+        if offset < 0 then
+            failwith "offset is less than zero"
+
+        if length <= 0 then
+            failwith "length is less than or equal to zero"
```

Comment:
More specific exception type: `ArgumentOutOfRangeException`

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/using-standard-exception-types#exception-and-systemexception